Documentation: add vmTools (#197247)
NixOS/nixpkgs: doc/builders/special/vm-tools.section.md

A bash script fragment that produces a disk image at `destination`.

### Attributes

* `size`. The disk size, in MiB.
* `fullName`. Name that will be written to `${destination}/nix-support/full-name`.
* `destination` (optional, default `$out`). Where to write the image files.

## `vmTools.runInLinuxVM` {#vm-tools-runInLinuxVM}

Run a derivation in a Linux virtual machine (using Qemu/KVM). By default, there is no disk image; the root filesystem is
a `tmpfs`, and the nix store is shared with the host (via the 9P protocol). Thus, any pure Nix derivation should run
unmodified.

If the build fails and Nix is run with the `-K` option, a script `run-vm` will be left behind in the temporary build
directory that allows you to boot into the VM and debug it interactively.

### Attributes

* `preVM` (optional). Shell command to be evaluated *before* the VM is started (i.e., on the host).
* `memSize` (optional, default `512`). The memory size of the VM in megabytes.
* `diskImage` (optional). A file system image to be attached to `/dev/sda` (Note that currently we expect the image to
  contain a filesystem, not a full disk image with a partition table etc).

### Examples

> ```nix
> # Build the derivation hello inside a VM
> runInLinuxVM hello
> 
> # Build inside a VM with extra memory
> runInLinuxVM (hello.overrideAttrs (_: { memSize = 1024; }))
> 
> # Use VM with a disk image (implicitly sets `diskImage`, see `vmTools.createEmptyImage`)
> runInLinuxVM (hello.overrideAttrs (_: { preVM = createEmptyImage { size = 1024; fullName = "vm-image"; }; }))
> ```

Reviewer: Do not use prose in comments, it's hard to read and to discern from *explanatory* comments as we're used to in actual code. 

Not sure about the scope aspect. While the samples are nice and easy on the eye, they're not self-contained and therefore not usable without contextual knowledge that we cannot assume from beginners (even if you say this is for advanced use cases - no one knows: that may be the first thing someone does with Nix, for whatever reason). 

If we had Sphinx, we could have full working samples and highlight the relevant part. I leave it up to you to decorate every one with `{ pkgs }: with pkgs;` or not.

````suggestion
Build the derivation hello inside a VM:

```nix
runInLinuxVM hello
```

Build inside a VM with extra memory:

```nix
runInLinuxVM (hello.overrideAttrs (_: { memSize = 1024; }))
```

Use VM with a disk image (implicitly sets `diskImage`, see 
[`vmTools.createEmptyImage`](#vm-tools-createEmptyImage)):

```nix
runInLinuxVM (hello.overrideAttrs (_: {
  preVM = createEmptyImage {
    size = 1024;
    fullName = "vm-image";
  };
}))
```
````

> <!-- TODO

Reviewer: (nit) We don't have a TODO/FIXME policy, but I have learned to keep FIXME's for posterity and TODO's for oneself. Therefore: finish or leave for posterity.

```suggestion
<!-- FIXME
```

## `vmTools.extractFs` {#vm-tools-extractFs}
## `vmTools.extractMTDfs` {#vm-tools-extractMTDfs}
-->

## `vmTools.runInLinuxImage` {#vm-tools-runInLinuxImage}

> Like [](#vm-tools-runInLinuxVM), but run the build not using the `stdenv` from the Nix store, but using the tools

Reviewer: ```suggestion
Like [](#vm-tools-runInLinuxVM), but instead of using `stdenv` from the Nix store, run the build using the tools
```

provided by `/bin`, `/usr/bin`, etc. from the specified filesystem image, which typically is a filesystem containing a

> non-NixOS Linux distribution.

Reviewer: ```suggestion
[FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard)-based Linux distribution.
```

[I think we should avoid the term *non-NixOS*.](https://discourse.nixos.org/t/should-we-give-a-name-to-nix-on-non-nixos/21020/3)

## `vmTools.makeImageTestScript` {#vm-tools-makeImageTestScript}

Generate a script that can be used to run an interactive session in the given image.

### Examples

> ```nix
> # Create a script for running a Fedora 27 VM
> makeImageTestScript diskImages.fedora27x86_64
> 
> # Create a script for running an Ubuntu 20.04 VM
> makeImageTestScript diskImages.ubuntu2004x86_64
> ```

Reviewer: As above.

````suggestion
Create a script for running a Fedora 27 VM:

```nix
makeImageTestScript diskImages.fedora27x86_64
```

Create a script for running an Ubuntu 20.04 VM:

```nix
makeImageTestScript diskImages.ubuntu2004x86_64
```
````

## `vmTools.diskImageFuns` {#vm-tools-diskImageFuns}

A set of functions that build a predefined set of minimal Linux distributions images.

### Images

* Fedora
  * `fedora26x86_64`
  * `fedora27x86_64`
* CentOS
  * `centos6i386`
  * `centos6x86_64`
  * `centos7x86_64`
* Ubuntu
  * `ubuntu1404i386`
  * `ubuntu1404x86_64`
  * `ubuntu1604i386`
  * `ubuntu1604x86_64`
  * `ubuntu1804i386`
  * `ubuntu1804x86_64`
  * `ubuntu2004i386`
  * `ubuntu2004x86_64`
* Debian
  * `debian9i386`
  * `debian9x86_64`
  * `debian10i386`
  * `debian10x86_64`

### Attributes

> * `size` (optional, defaults to `4096`). The size of the image, in megabytes.

Reviewer: As above.

```suggestion
* `size` (optional, defaults to `4096`). The size of the image, in MiB.
```

> * `extraPackages` (optional). A list names of additional packages from the distribution that should be included in the
>   image.

Reviewer: Please don't wrap lines arbitrarily. I know, this is a contentious topic. The documentation team has agreed to roll out a [convention to break lines at sentence boundaries](https://github.com/NixOS/nix.dev/blob/master/CONTRIBUTING.md?plain=1#L175-L185) for the simple reason that it's easier to edit and diff.

```suggestion
* `extraPackages` (optional). A list names of additional packages from the distribution that should be included in the image.
```

Not super important here, and I won't ask you to reformat everything, just to make you aware.

### Examples

> ```nix
> # 8G image containing Firefox in addition to the default packages.
> diskImageFuns.ubuntu2004x86_64 { extraPackages = [ "firefox" ]; size = 8192; }
> ```

Reviewer: ````suggestion
8GiB image containing Firefox in addition to the default packages:

```nix
diskImageFuns.ubuntu2004x86_64 { extraPackages = [ "firefox" ]; size = 8192; }
```
````
